core/store/src/trie/prefetching_trie_storage.rs

```diff
@@ -0,0 +1,402 @@
+use crate::trie::POISONED_LOCK_ERR;
+use crate::{DBCol, StorageError, Store, TrieCache, TrieCachingStorage, TrieStorage};
+use near_primitives::hash::CryptoHash;
+use near_primitives::shard_layout::ShardUId;
+use near_primitives::trie_key::TrieKey;
+use near_primitives::types::TrieNodesCount;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tracing::error;
+
+const MAX_QUEUED_WORK_ITEMS: usize = 16 * 1024;
+const MAX_PREFETCH_STAGING_MEMORY: usize = 200 * 1024 * 1024;
+/// How much memory capacity is reserved for each prefetch request.
+/// Set to 4MiB, the same as `max_length_storage_value`.
+const PREFETCH_RESERVED_BYTES_PER_SLOT: usize = 4 * 1024 * 1024;
```

Comment:
It means that if 50 slots are occupied, we can't insert new prefetch requests anymore. One receipt can include a batch of 100 storage reads, which itself may trigger, say 20 nodes read, so such scenario would require 2K slots.
Is prefetcher fast enough to keep the queue not full? If not, can we call `Trie::get_ref` instead of `Trie::get` - so we could reduce value size limit to 1K - node size limit?

Reply:
I should clarify this in the comment: This is the reserved bytes before we start fetching, at which point we assume the worst case. But once we have the value, we will use the actual size. So with 8 IO threads, we only charge this pessimistic value for 8 slots at a time. With that in mind, the prefetcher is fast enough.
If we wanted to go for many more threads, I think your idea would be perfect to address this issue. But as things stand, I tend towards prefetching everything.